cpu/esp32/esp-now/esp_now_netif.c

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (C) 2018 Timo Rothenpieler
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_esp32_esp_now
+ * @{
+ *
+ * @file
+ * @brief       Netif interface for the ESP-NOW WiFi P2P protocol
+ *
+ * @author Timo Rothenpieler <timo.rothenpieler@uni-bremen.de>
+ */
+
+#include <assert.h>
+#include <stdlib.h>
+
+#include <sys/uio.h>
+
+#include "net/netdev.h"
+#include "net/gnrc.h"
+#include "esp_now_params.h"
+#include "esp_now_netdev.h"
+#include "esp_now_netif.h"
+#include "net/gnrc/netif.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static int _send(gnrc_netif_t *netif, gnrc_pktsnip_t *pkt)
+{
+    uint8_t mac[ESP_NOW_ADDR_LEN];
+    uint8_t plbuf[ESP_NOW_MAX_SIZE_RAW];
+    netdev_t *dev = netif->dev;
+
+    assert(pkt != NULL);
+
+    gnrc_netif_hdr_t *netif_hdr;
+    gnrc_pktsnip_t *payload;
+
+    payload = pkt->next;
+
+    if (pkt->type != GNRC_NETTYPE_NETIF) {
+        DEBUG("gnrc_esp_now: First header was not generic netif header\n");
+        gnrc_pktbuf_release(pkt);
+        return -EBADMSG;
+    }
+
+    netif_hdr = (gnrc_netif_hdr_t*)pkt->data;
+
+    if (netif_hdr->flags & (GNRC_NETIF_HDR_FLAGS_BROADCAST | GNRC_NETIF_HDR_FLAGS_MULTICAST)) {
+        /* ESP-NOW does not support multicast, always broadcast */
+        memset(mac, 0xff, ESP_NOW_ADDR_LEN);
+    } else if (netif_hdr->dst_l2addr_len == ESP_NOW_ADDR_LEN) {
+        memcpy(mac, gnrc_netif_hdr_get_dst_addr(netif_hdr), ESP_NOW_ADDR_LEN);
+    } else {
+        DEBUG("gnrc_esp_now: destination address had unexpected format"
+              "(flags=%d, dst_l2addr_len=%d)\n", netif_hdr->flags, netif_hdr->dst_l2addr_len);
+        gnrc_pktbuf_release(pkt);
+        return -EBADMSG;
+    }
+
+    iolist_t payload_iolist = {
+        .iol_base = plbuf,
+        .iol_len = sizeof(esp_now_pkt_hdr_t)
+    };
+
+    iolist_t iolist = {
+        .iol_base = mac,
+        .iol_len = sizeof(mac),
+        .iol_next = &payload_iolist
+    };
+
+    esp_now_pkt_hdr_t *esp_hdr = payload_iolist.iol_base;
+
+    switch (payload->type) {
+#ifdef MODULE_GNRC_SIXLOWPAN
+        case GNRC_NETTYPE_SIXLOWPAN:
+            esp_hdr->flags = 1;
```

Comment:
Some some defines instead of magic numbers would be helpful to understand what the flags actually mean.

Reply:
Using a define in latest version now.